File: jbake-core/src/main/java/org/jbake/render/DocumentsRenderer.java
package org.jbake.render;

import org.apache.commons.configuration2.CompositeConfiguration;
import org.jbake.app.ContentStore;
import org.jbake.app.DocumentList;
import org.jbake.app.Renderer;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.exception.RenderingException;
import org.jbake.model.DocumentModel;
import org.jbake.model.ModelAttributes;

import java.io.File;
import java.util.LinkedList;
import java.util.List;

public class DocumentsRenderer implements RenderingTool {

    @Override
    public int render(Renderer renderer, ContentStore db, JBakeConfiguration config) throws RenderingException {
        int renderedCount = 0;
        final List<String> errors = new LinkedList<>();

        DocumentList<DocumentModel> documentList = db.getUnrenderedContent();
        for (DocumentModel document : documentList) {
            try {
                DocumentList<DocumentModel> typedDocList = db.getAllContent(document.getType());
                DocumentModel prev = getPrevDoc(typedDocList, document);
                DocumentModel next = getNextDoc(typedDocList, document);
                document.setPreviousContent(prev);
                document.setNextContent(next);

                renderer.render(document);
                db.markContentAsRendered(document);
                renderedCount++;

            } catch (Exception e) {
                errors.add(e.getMessage());
            }
        }

        if (!errors.isEmpty()) {
            StringBuilder sb = new StringBuilder();
            sb.append("Failed to render documents. Cause(s):");
            for (String error : errors) {
                sb.append("\n").append(error);
            }
            throw new RenderingException(sb.toString());
        } else {
            return renderedCount;
        }
    }

    private DocumentModel getNextDoc(DocumentList<DocumentModel> typedList, DocumentModel doc) {
        int typedListIndex = typedList.indexOf(doc);
        if (typedList.getFirst().equals(doc)) {
            // initial doc in typed list so there is no next
            return null;
        } else {
            while (true) {
                try {
                    DocumentModel nextDoc = typedList.get(typedListIndex - 1);
                    if (isPublished(nextDoc)) {
                        return getContentForNav(nextDoc);
                    } else {
                        typedListIndex--;
                    }
                } catch (IndexOutOfBoundsException ex) {
                    return null;
                }
            }
        }
    }

    private DocumentModel getPrevDoc(DocumentList<DocumentModel> typedList, DocumentModel doc) {
        int typedListIndex = typedList.indexOf(doc);
        if (typedList.getLast().equals(doc)) {
            // last doc in typed list so there is no previous
            return null;
        } else {
            while (true) {
                try {
                    DocumentModel prevDoc = typedList.get(typedListIndex + 1);
                    if (isPublished(prevDoc)) {
                        return getContentForNav(prevDoc);
                    } else {
                        typedListIndex++;
                    }
                } catch (IndexOutOfBoundsException ex) {
                    return null;
                }
            }
        }
    }

    private boolean isPublished(DocumentModel document) {
        // Attributes.Status.PUBLISHED_DATE cannot occur here
        // because it's converted TO either PUBLISHED or DRAFT in the Crawler.
        return ModelAttributes.Status.PUBLISHED.equals(document.getStatus());
    }

    /**
     * Creates a simple content model to use in individual post navigations.
     *
     * @param document original
     * @return navigation model for the 'document'
     */
    private DocumentModel getContentForNav(DocumentModel document) {
        DocumentModel navDocument = new DocumentModel();
        navDocument.setNoExtensionUri(document.getNoExtensionUri());
        navDocument.setUri(document.getUri());
        navDocument.setTitle(document.getTitle());
        return navDocument;
    }

    @Override
    public int render(Renderer renderer, ContentStore db, File destination, File templatesPath, CompositeConfiguration config) throws RenderingException {
        return render(renderer, db, null);
    }
}

File: jbake-core/src/main/java/org/jbake/render/DocumentsRenderer.kt
package org.jbake.render

import org.jbake.app.ContentStore
import org.jbake.app.DocumentList
import org.jbake.app.Renderer
import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.model.DocumentModel
import org.jbake.model.ModelAttributes
import org.jbake.template.RenderingException
import org.jbake.util.Logging.logger
import org.slf4j.Logger
import java.util.*

class DocumentsRenderer : RenderingTool {
    private val log: Logger by logger()

    @Throws(RenderingException::class)
    @Deprecated("Overrides deprecated member")
    override fun render(renderer: Renderer, db: ContentStore, config: JBakeConfiguration): Int {
        var renderedCount = 0
        val errors: MutableList<String> = LinkedList<String>()

        val documentList = db.unrenderedContent
        for (document in documentList) {
            try {
                val typedDocList = db.getAllContent(document.type)
                val prev = getPrevDoc(typedDocList, document)
                val next = getNextDoc(typedDocList, document)
                document.previousContent = prev
                document.nextContent = next

                renderer.render(document)
                db.markContentAsRendered(document)
                renderedCount++
            } catch (e: Exception) {
                // Log full stacktrace to help debugging in tests
                log.error("Error rendering document {} (type={})", document.name, document.type, e)
                errors.add(e.message ?: (e.toString() + " " + e.stackTrace.first().toString()))
            }
        }

        if (!errors.isEmpty()) {
            val sb = StringBuilder()
            sb.append("Failed to render documents. Cause(s):")
            for (error in errors) {
                sb.append("\n").append(error)
            }
            throw RenderingException(sb.toString())
        }
        return renderedCount
    }

    private fun getNextDoc(typedList: DocumentList<DocumentModel>, doc: DocumentModel?): DocumentModel? {
        var typedListIndex = typedList.indexOf(doc)
        if (typedList.first() == doc) {
            // initial doc in typed list so there is no next
            return null
        }
        while (true) {
            try {
                val nextDoc = typedList[typedListIndex - 1]
                if (isPublished(nextDoc)) {
                    return getContentForNav(nextDoc)
                }
                typedListIndex--
            } catch (ex: IndexOutOfBoundsException) {
                return null
            }
        }
    }

    private fun getPrevDoc(typedList: DocumentList<DocumentModel>, doc: DocumentModel?): DocumentModel? {
        var typedListIndex = typedList.indexOf(doc)
        if (typedList.last() == doc) {
            // last doc in typed list so there is no previous
            return null
        }
        while (true) {
            try {
                val prevDoc = typedList[typedListIndex + 1]
                if (isPublished(prevDoc)) {
                    return getContentForNav(prevDoc)
                }
                typedListIndex++
            } catch (ex: IndexOutOfBoundsException) {
                return null
            }
        }
    }

    private fun isPublished(document: DocumentModel): Boolean {
        // Attributes.Status.PUBLISHED_DATE cannot occur here
        // because it's converted TO either PUBLISHED or DRAFT in the Crawler.
        return ModelAttributes.Status.PUBLISHED == document.status
    }

    /**
     * Creates a simple content model to use in individual post navigations.
     *
     * @param document original
     * @return navigation model for the 'document'
     */
    private fun getContentForNav(document: DocumentModel): DocumentModel {
        val navDocument = DocumentModel()
        navDocument.noExtensionUri = document.noExtensionUri
        navDocument.uri = document.uri
        navDocument.title = document.title
        return navDocument
    }
}
